# Introduction
In this project, I am going to build a spam filter for SMS messages.

To classify messages as spam or non-spam, the computer:

1. Has to learn how humans classify messages.
2. Use that human knowledge to estimate probabilities for new messages — probabilities for spam and non-spam.
3. Classifies a new message based on these probability values — if the probability for spam is greater, then it classifies the message as spam. Otherwise, it classifies it as non-spam (if the two probability values are equal, then we may need a human to classify the message).

The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from the The [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). The data collection process is described in more detail on this [webpage](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).

# Exploring the Dataset

In [1]:
import pandas as pd
sms_spam = pd.read_csv('SMSSpamCollection', sep = '\t', header=None, names = ['Label', 'SMS'])
print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [2]:
sms_spam['Label'].value_counts(normalize = True)

ham     0.865937
spam    0.134063
Name: Label, dtype: float64

About 87% of the messages are ham, and the remaining 13% are spam. This sample looks representative, as in real life also most messages that people receive are ham.

# Training and Test Set
After randoming the data rows, to ensure the spam and ham(non-spam) messages are evenly distributed in the set, the 80% of the data will be divided into training set and the remaining in test set.

In [3]:
randomized_set = sms_spam.sample(frac = 1, random_state = 1)
training_set_index = round(len(randomized_set)*0.80)
train_set = randomized_set[:training_set_index].reset_index(drop = True)
test_set = randomized_set[training_set_index:].reset_index(drop = True)

Analyzing the percentage of spam and ham messages in the training and test sets. the expectation is that the percentages will be close to what was in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

In [4]:
train_set['Label'].value_counts(normalize = True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [5]:
test_set['Label'].value_counts(normalize = True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

The result is in allignment with the full dataset.

# Cleaning the Data
In this step all punctuations will be removed and the words will be converted to lower case so that, Secret and secret will be one word. 

In [6]:
train_set['SMS'] = train_set['SMS'].str.replace('\W', " ")
train_set['SMS'] = train_set['SMS'].str.lower()
test_set['SMS'] = test_set['SMS'].str.replace('\W', " ")
test_set['SMS'] = test_set['SMS'].str.lower()
train_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

## Creating a Vocabulary

In [7]:
train_set['SMS']= train_set['SMS'].str.split()
vocabulary = []
for sms in train_set['SMS']:
     for word in sms:
        vocabulary.append(word)
        vocabulary = list(set(vocabulary))

len(vocabulary)

7783

In [8]:
word_counts_per_sms = {unique_word: [0] * len(train_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(train_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1
        
word_counts = pd.DataFrame(word_counts_per_sms)
training_set_clean = pd.concat([train_set, word_counts], axis=1)
training_set_clean.head()

Label                                                SMS  ni8  08709222922  \
0   ham                  [yep, by, the, pretty, sculpture]    0            0   
1   ham  [yes, princess, are, you, going, to, make, me,...    0            0   
2   ham                    [welp, apparently, he, retired]    0            0   
3   ham                                           [havent]    0            0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...    0            0   

   goodtime  started  location  jane  amazing  4d  ...    450pw  sell  amma  \
0         0        0         0     0        0   0  ...        0     0     0   
1         0        0         0     0        0   0  ...        0     0     0   
2         0        0         0     0        0   0  ...        0     0     0   
3         0        0         0     0        0   0  ...        0     0     0   
4         0        0         0     0        0   0  ...        0     0     0   

   christians  whens  concerned  pics  systems  smiling  80182  
0           0      0          0     0        0        0      0  
1           0      0          0     0        0        0      0  
2           0      0          0     0        0        0      0  
3           0      0          0     0        0        0      0  
4           0      0          0     0        0        0      0  

[5 rows x 7785 columns]

# Calculating Constants First

The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

$$
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam) \\\
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
$$
Also, to calculate P(wi|Spam) and P(wi|Ham) inside the formulas above, we'll need to use these equations:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}} \\\
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$
Some of the terms in the four equations above will have the same value for every new message. We can calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, the training set is used to calculate:

P(Spam) and P(Ham)
NSpam, NHam, NVocabulary
Laplace smoothing and set $\alpha = 1$ is also used.

In [9]:
# Isolating spam and ham messages first
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

# Calculating Parameter
Now that we have the constant terms calculated above, we can move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

In [10]:
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

# Classifying a New Message

Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:

Takes in as input a new message (w1, w2, ..., wn).
Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:<br>
If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham. <br>
If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam. <br>
If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.

In [11]:
import re

def classify(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()

    #calculations  
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)

    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [12]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')


P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [13]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


# Measuring the Spam Filter's Accuracy
The two results above look promising, but let's see how well the filter does on our test set, which has 1,114 messages.

We'll start by writing a function that returns classification labels instead of printing them.

In [14]:
def classify_test_set(message):    
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

In [15]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          later i guess  i needa do mcat study too        ham
1   ham             but i haf enuff space got like 4 mb          ham
2  spam  had your mobile 10 mths  update to latest oran...      spam
3   ham  all sounds good  fingers   makes it difficult ...       ham
4   ham  all done  all handed in  don t know if mega sh...       ham

Now, we'll write a function to measure the accuracy of our spam filter to find out how well our spam filter does.

In [17]:
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833


The accuracy is close to 98.74%, which is really good. Our spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,100 correctly.

# Next Steps
In this project, we managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set we used, which is a pretty good result. 

Next steps include:

Analyze the 14 messages that were classified incorrectly and try to figure out why the algorithm classified them incorrectly<br>
Make the filtering process more complex by making the algorithm sensitive to letter case